apps/emqx_plugin_libs/src/emqx_placeholder.erl

```diff
@@ -132,10 +236,42 @@ get_phld_var(Fun, Data) when is_function(Fun) ->
 get_phld_var(Phld, Data) ->
     emqx_rule_maps:nested_get(Phld, Data).
 
-replace_with(Tmpl, '?') ->
-    re:replace(Tmpl, ?EX_PLACE_HOLDER, "?", [{return, binary}, global]);
-replace_with(Tmpl, '$n') ->
-    Parts = re:split(Tmpl, ?EX_PLACE_HOLDER, [{return, binary}, trim, group]),
+preproc_var_re(#{placeholders := PHs}) ->
+    "(" ++ string:join([ph_to_re(PH) || PH <- PHs], "|") ++ ")";
```

Comment:
Nit: it's not necessary to glue the strings together and return iodata() instead. `re:run` supports iolist as a RE argument: https://www.erlang.org/doc/man/re.html#run-3